File: python/qsort.py
import random

def quicksort(array):
  if(len(array) < 2):
    return array
  else:
    pivot = array.pop(get_pivot(array))
    less = []
    greater = []
    for x in array:
      if(x <= pivot):
        less.append(x)
      else:
        greater.append(x)

    qsort_less = quicksort(less)
    qsort_less.extend([pivot])
    qsort_less.extend(quicksort(greater))
    return qsort_less

def get_pivot(array):
  return random.randrange(0, len(array))



print quicksort([5,2,1,12,3,5,61,8])

File: javascript/qsort.js
(function(){

  Quicksort = function(){

  };

  Quicksort.prototype.sort = function(array) {
    if (array.length < 2) {
      return array;
    }

    var greater = [];
    var less = [];
    var pivot = get_pivot(array);

    for(var i = 0; i < array.length; i++ ){
      if (array[i] < pivot) {
        less.push(array[i]);
      }
      else {
        greater.push(array[i]);
      }
    }

    return merge(this.sort(less), this.sort(greater));
  };

  var merge = function(array1, array2) {
    return array1.concat(array2);
  };

  var get_pivot = function(array) {
    return array[parseInt((Math.random() * 100) % array.length)];
  };

})();
